src/v/datalake/batching_parquet_writer.cc

```diff
@@ -0,0 +1,83 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
+ *
+ * Licensed as a Redpanda Enterprise file under the Redpanda Community
+ * License (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * https://github.com/redpanda-data/redpanda/blob/master/licenses/rcl.md
+ */
+
+#include "datalake/batching_parquet_writer.h"
+
+#include "bytes/iostream.h"
+#include "datalake/arrow_translator.h"
+#include "datalake/data_writer_interface.h"
+
+#include <seastar/core/file-types.hh>
+#include <seastar/core/fstream.hh>
+#include <seastar/core/future.hh>
+#include <seastar/core/iostream.hh>
+#include <seastar/core/seastar.hh>
+
+#include <cstdint>
+
+namespace datalake {
+
+batching_parquet_writer::batching_parquet_writer(
+  iceberg::struct_type schema,
+  size_t row_count_threshold,
+  size_t byte_count_threshold)
+  : _iceberg_to_arrow(schema)
+  , _arrow_to_iobuf(_iceberg_to_arrow.build_arrow_schema())
+  , _row_count_threshold{row_count_threshold}
+  , _byte_count_threshold{byte_count_threshold} {}
+
+ss::future<>
+batching_parquet_writer::initialize(std::filesystem::path output_file_path) {
+    _output_file = co_await ss::open_file_dma(
+      output_file_path.string(),
+      ss::open_flags::create | ss::open_flags::truncate | ss::open_flags::wo);
+
+    _output_stream = co_await ss::make_file_output_stream(_output_file);
+}
+
+ss::future<void> batching_parquet_writer::add_data_struct(
+  iceberg::value data, int64_t approx_size) {
+    _iceberg_to_arrow.add_data(std::move(data));
+    _row_count++;
+    _byte_count += approx_size;
+
+    if (
+      _row_count >= _row_count_threshold
+      || _byte_count > _byte_count_threshold) {
+        co_await write_row_group();
+    }
```

Comment:
Just a thought, if this is meant to be a building block for higher level abstractions, wondering if we should just expose the row and byte count and let the flush policy be taken care of at a higher level